wasm/src/varint.rs

```diff
@@ -0,0 +1,148 @@
+use crate::error::Error;
+
+const SEGMENT_BITS: u8 = 0x7F;
+const CONTINUE_BIT: u8 = 0x80;
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct DecodedResult<T> {
+    pub value: T,
+    pub bytes_read: u8,
+}
+
+pub fn decode_u32<R: std::io::Read>(reader: &mut R) -> Result<DecodedResult<u32>, Error> {
+    let mut length = 0;
+    let mut value = 0;
+
+    loop {
+        let mut bytes: [u8; 1] = [0; 1];
+        reader.read_exact(&mut bytes)?;
+        value |= ((bytes[0] & SEGMENT_BITS) as u32) << length;
+
+        length += 7;
+        if bytes[0] & CONTINUE_BIT == 0 {
+            break;
+        }
+
+        if length > 32 {
+            return Err(Error::TooManyBytes {
+                expected: 5,
+                found: length / 7 + 1,
+            });
+        }
+    }
+
+    length += length / 7;
+
+    Ok(DecodedResult {
+        value,
+        bytes_read: length / 8,
+    })
+}
+
+pub fn decode_u64<R: std::io::Read>(reader: &mut R) -> Result<DecodedResult<u64>, Error> {
+    let mut length = 0;
+    let mut value = 0;
+
+    loop {
+        let mut bytes: [u8; 1] = [0; 1];
+        reader.read_exact(&mut bytes)?;
+        value |= ((bytes[0] & SEGMENT_BITS) as u64) << length;
+
+        length += 7;
+        if bytes[0] & CONTINUE_BIT == 0 {
+            break;
+        }
+
+        if length > 64 {
```

Comment:
Wait, would this overflow for `length == 64`?

Reply:
Oh shoot you are right...
Should've been `>=` not `>`

Reply:
Opened a pr for that